kernel/src/log_segment/mod.rs

```diff
@@ -0,0 +1,374 @@
+    DeltaResult, Engine, EngineData, Error, Expression, ExpressionRef, FileSystemClient, Version,
+};
+use itertools::Itertools;
+use std::cmp::Ordering;
+use std::convert::identity;
+use std::sync::{Arc, LazyLock};
+use tracing::warn;
+use url::Url;
+
+#[cfg(test)]
+mod tests;
+
+/// A [`LogSegment`] represents a contiguous section of the log and is made of checkpoint files
+/// and commit files and guarantees the following:
+///     1. Commit file versions will not have any gaps between them.
+///     2. If checkpoint(s) is/are present in the range, only commits with versions greater than the most
+///        recent checkpoint version are retained. There will not be a gap between the checkpoint
+///        version and the first commit version.
+///     3. All checkpoint_parts must belong to the same checkpoint version, and must form a complete
+///        version. Multi-part checkpoints must have all their parts.
+///
+/// [`LogSegment`] is used in [`Snapshot`] when built with [`LogSegment::for_snapshot`], and
+/// and in `TableChanges` when built with [`LogSegment::for_table_changes`].
+///
+/// [`Snapshot`]: crate::snapshot::Snapshot
+#[derive(Debug)]
+#[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
+pub(crate) struct LogSegment {
+    pub end_version: Version,
+    pub log_root: Url,
+    /// Sorted commit files in the log segment (ascending)
+    pub sorted_commit_files: Vec<ParsedLogPath>,
+    /// Checkpoint files in the log segment.
+    pub checkpoint_parts: Vec<ParsedLogPath>,
+}
+
+impl LogSegment {
+    fn try_new(
+        sorted_commit_files: Vec<ParsedLogPath>,
+        checkpoint_parts: Vec<ParsedLogPath>,
+        log_root: Url,
+        end_version: Option<Version>,
+    ) -> DeltaResult<Self> {
+        // We require that commits that are contiguous. In other words, there must be no gap between commit versions.
+        require!(
+            sorted_commit_files
+                .windows(2)
+                .all(|cfs| cfs[0].version + 1 == cfs[1].version),
+            Error::generic(format!(
+                "Expected ordered contiguous commit files {:?}",
+                sorted_commit_files
+            ))
+        );
+
+        // There must be no gap between a checkpoint and the first commit version. Note that
+        // that all checkpoint parts share the same version.
+        if let (Some(checkpoint_file), Some(commit_file)) =
+            (checkpoint_parts.first(), sorted_commit_files.first())
+        {
+            require!(checkpoint_file.version + 1 == commit_file.version,
+            Error::generic(format!("Expected commit file version {} to be next version to checkpoint file version {}", commit_file.version, checkpoint_file.version )))
+        }
+
+        // Get the effective version from chosen files
+        let version_eff = sorted_commit_files
+            .last()
+            .or(checkpoint_parts.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
```

Comment:
Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/495#discussion_r1843161719
> Would it make more sense to use an Error::generic here, or create an new error variant?

Generic seems fine for now. 

We may want a tracking issue to make a new error class to cover all missing file situations we might encounter during log listing?